cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,469 @@
+          source_scalar, scatter_iter, scatter_rows, target_col, mr, stream);
+      });
+
+    scatter_scalar_bitmask(source, scatter_iter, scatter_rows, result, mr, stream);
+
+    return std::make_unique<table>(std::move(result));
+  }
+
+  template <typename T, std::enable_if_t<not std::is_integral<T>::value
+      or std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      std::vector<std::unique_ptr<scalar>> const& source,
+      column_view const& indices, table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_FAIL("Scatter index column must be an integral, non-boolean type");
+  }
+};
+
+struct scatter_to_tables_impl {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value
+      and not std::is_same<T, bool8>::value>* = nullptr>
+  std::vector<std::unique_ptr<table>> operator()(
+      table_view const& input, column_view const& partition_map,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    // Make a mutable copy of the partition map
+    auto d_partitions = rmm::device_vector<T>(
+      partition_map.begin<T>(), partition_map.end<T>());
+
+    // Initialize gather maps and offsets to sequence
+    auto d_gather_maps = rmm::device_vector<size_type>(partition_map.size());
+    auto d_offsets = rmm::device_vector<size_type>(partition_map.size());
+    thrust::sequence(rmm::exec_policy(stream)->on(stream),
+      d_gather_maps.begin(), d_gather_maps.end());
+    thrust::sequence(rmm::exec_policy(stream)->on(stream),
+      d_offsets.begin(), d_offsets.end());
+
+    // Sort sequence using partition map as key to generate gather maps
+    thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+      d_partitions.begin(), d_partitions.end(), d_gather_maps.begin());
+
+    // Reduce unique partitions to extract gather map offsets from sequence
+    auto end = thrust::unique_by_key(rmm::exec_policy(stream)->on(stream),
+      d_partitions.begin(), d_partitions.end(), d_offsets.begin());
+
+    // Copy partition indices and gather map offsets to host
+    auto partitions = thrust::host_vector<T>(d_partitions.begin(), end.first);
+    auto offsets = thrust::host_vector<size_type>(d_offsets.begin(), end.second);
+    offsets.push_back(partition_map.size());
+
+    CUDF_EXPECTS(partitions.front() >= 0, "Invalid negative partition index");
+    auto output = std::vector<std::unique_ptr<table>>(partitions.back() + 1);
+
+    size_t next_partition = 0;
+    for (size_t index = 0; index < partitions.size(); ++index) {
+      auto const partition = static_cast<size_t>(partitions[index]);
+
+      // Create empty tables for unused partitions
+      for (; next_partition < partition; ++next_partition) {
+        output[next_partition] = empty_like(input, stream);
+      }
+
+      // Gather input rows for the current partition
+      auto const data = d_gather_maps.data().get() + offsets[index];
```

Comment:
Hmmm, couldn't the gather map be a zero-copy slice view of `d_partitions`? Then you would only have to read back the offsets to host, and re-copy the map data to device for every partition. Also, rather than the loop to create empty tables, couldn't you just detect that `size` is zero? 

e.g. 

```
if (size > 0)
  output[partition] = gather(input, gather_map, false, false, false, mr, stream);
else
  output[partition] = empty_like(input, stream);
```

Reply:
I don't follow. I have 3 buffers: unique partitions (only those that appear in the partition map, so for e.g 2, 4, 5 would mean that 0, 1, 3 get empty tables), gather map offsets (same size as unique partitions + 1), and the full-size gather map. I copy the first two to host, but the gather map itself stays on device.
The problem with checking for size is that partitions/offsets that don't get written to aren't present in the array, so I instead determine where to put empty tables by which indices _don't_ appear in partitions

Reply:
I see. Didn't realize the difference. Was hoping to avoid as much round trip as  possible.